### PR TITLE
GitIgnore generate improvements

### PR DESCRIPTION
**Pull request by yvesmendes**
Adjusment the default constructor of GitIgnore to initialize the default values of GitIgnoresections

```
public GitIgnore() {
		this.ignoreSections = initializeGitIgnoreSections();
	}
	private Map<GitEnum, GitIgnoreSection> initializeGitIgnoreSections() {
		Map<GitEnum, GitIgnoreSection> ignoreSections = new HashMap<>();
		Stream.of(GitEnum.values()).forEach((gitEnum) -> {
			GitIgnoreSection gitIgnoreSection = new GitIgnoreSection(gitEnum.getName());
			gitIgnoreSection.add(gitEnum.getIgnores());
			ignoreSections.put(gitEnum, gitIgnoreSection);
		});
		return ignoreSections;
	}
```
Create a GitEnum with the default values of GitIgnoreSection

```
public enum GitEnum {

	/**
	 * STS value of GitignoreSection.
	 */
	STS("STS", new String[] { ".apt_generated", ".classpath", ".factorypath", ".project", ".settings", ".springBeans",
			".sts4-cache" }),
	/**
	 * Intellij IDEA value of GitignoreSection.
	 */
	INTELLIJ_IDEA("IntelliJ IDEA", new String[] { ".idea", "*.iws", "*.iml", "*.ipr" }),
	/**
	 * Netbeans value of GitignoreSection.
	 */
	NET_BEANS("NetBeans", new String[] { "/nbproject/private/", "/nbbuild/", "/dist/", "/nbdist/", "/.nb-gradle/" }),
	/**
	 * VS Code value of GitignoreSection.
	 */
	VS_CODE("VS Code", new String[] { ".vscode/" }),
	/**
	 * General value of GitignoreSection.
	 */
	GENERAL(null, null);

	/**
	 * List of ignore values.
	 */
	String[] ignores;

	/**
	 * Description name of gitignore section.
	 */
	String name;

	GitEnum(String name, String[] asList) {
		this.ignores = asList;
		this.name = name;
	}

	public String[] getIgnores() {
		return this.ignores;
	}

	public String getName() {
		return this.name;
	}
}
```

Refactoring all specific methods this.gitIgnore.getXXX() to this.getGitIgnoreSection(GitEnum.GENERAL) (or equivalent), turning the code clear and easy to extend (OCP principle).

### CODE REVIEW

File: initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitEnum.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * limitations under the License.
+ */
+
+package io.spring.initializr.generator.spring.scm.git;
+
+/**
+ * Default configurations for gitignore file.
+ * @author Yves Galvão
+ */
+public enum GitEnum {
```
Comment (snicoll) [inline initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitEnum.java:23]: I am not sure why we'd want to move this to an enum. Can you expand why you think this improves the status quo?

File: initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitIgnore.java [hunk 2]
```diff
@@ -28,51 +30,37 @@
...
  *
  * @author Stephane Nicoll
+ * @author Yves Galvão
  */
 public class GitIgnore {
 
-	private final GitIgnoreSection general = new GitIgnoreSection(null);
+	private final Map<GitEnum, GitIgnoreSection> ignoreSections;
 
-	private final GitIgnoreSection sts = new GitIgnoreSection("STS");
```
Comment (snicoll) [inline initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitIgnore.java:36]: The idea of this is to model the `.gitgnore` files with well-defined sections. If you don't set them, they are not rendered (as the class javadoc states). I don't understand why moving to a map based approach is better. Can you expand on that?

**Review by snicoll — COMMENTED** [review]
Thanks for the PR! I've added a couple comments as I'd like to better understand the change.

File: initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitEnum.java (hunk 1 above)
Comment (yvesmendes) [reply initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitEnum.java:23]: Hi @snicoll, I made some customs in the code for the company where I work, and looking for SOLID principles, I thinking that's flow is violated the [Open-closed principle](https://en.wikipedia.org/wiki/Open–closed_principle). If another **GitIgnoreSection** is needed on the flow, the developer should by modify the **GitIgnore class**, the **write method on GitIgnore class**, the **createGitIgnore method on GitProjectGenerationConfiguration**, and **add a custom method to return the list of ignores on the GitIgnore class**. In contrast, with this refactoring the only one adjustment is to **add a new value on the created GitEnum**.

File: initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitIgnore.java (hunk 2 above)
Comment (yvesmendes) [reply initializr-generator-spring/src/main/java/io/spring/initializr/generator/spring/scm/git/GitIgnore.java:36]: It's not the write method on GitIgnore thas is responsible for render the GitIgnoreSection informations on .gitignore file? If not, maybe I have a misinterpretation of the flow, and this PR does not make sense.

**Comment by snicoll** [comment]
I am going to close this now as I don't think we want to extend how `.gitignore` is generated this way. I recognize that the current state of affairs can be improved though so I've created #1426 to investigate. Thanks again for the PR.

**Pull request closed without merging**